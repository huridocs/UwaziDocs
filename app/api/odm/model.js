import mongoose from 'mongoose';

const generateID = mongoose.Types.ObjectId;
export {generateID};

export default (MongooseModel) => {
  return {
    db: MongooseModel,
    save: (data) => {
      if (Array.isArray(data)) {
        let promises = data.map((entry) => {
          if (entry._id) {
            return MongooseModel.findOneAndUpdate({_id: entry._id}, entry, {new: true, lean: true});
          }
          return MongooseModel.create(entry).then(saved => saved.toObject());
        });
        return Promise.all(promises);
      }

      if (data._id) {
        return MongooseModel.findOneAndUpdate({_id: data._id}, data, {new: true, lean: true});
      }
      return MongooseModel.create(data).then(saved => saved.toObject());
    },

    get: (query, select = '', pagination = {}) => {
      return MongooseModel.find(query, select, Object.assign({lean: true}, pagination));
    },

    count: (condition) => {
      return MongooseModel.count(condition);
    },

    getById: (id) => {
      return MongooseModel.findById(id, {}, {lean: true});
    },

    delete: (condition) => {
      let cond = condition;
      if (mongoose.Types.ObjectId.isValid(condition)) {
        cond = {_id: condition};
      }
      return MongooseModel.remove(cond);
    }
  };
};
