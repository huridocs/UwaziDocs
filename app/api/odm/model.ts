import { WithId as _WithId } from 'mongodb';
import mongoose, {
  Schema,
  UpdateQuery,
  ModelUpdateOptions,
  FilterQuery,
  QueryOptions,
} from 'mongoose';
import { ObjectIdSchema } from 'shared/types/commonTypes';
import { MultiTenantMongooseModel } from './MultiTenantMongooseModel';
import { createUpdateLogHelper, UpdateLogger } from './logHelper';

/** Ideas!
 *  T is the actual model-specific document Schema!
 *  DataType should be the specific Schema with Document, in order to have _id and other Document specific characteristcs
 *  WithId should be returned by get model?
 *  Do we need to type Uwazi specific Filter, Update and Query?
 */

export type DataType<T> = T & { _id?: ObjectIdSchema };

export type WithId<T> = _WithId<T>;

export type EnforcedWithId<T> = WithId<mongoose.EnforceDocument<DataType<T>, {}>>;

export type UwaziFilterQuery<T> = FilterQuery<T>;
export type UwaziUpdateQuery<T> = UpdateQuery<DataType<T>>;
export type UwaziQueryOptions = QueryOptions;

const generateID = mongoose.Types.ObjectId;
export { generateID };

export class OdmModel<T> {
  db: MultiTenantMongooseModel<T>;

  logHelper: UpdateLogger<T>;

  private documentExists(data: Partial<DataType<T>>) {
    return this.db.findById(data._id, '_id');
  }

  constructor(logHelper: UpdateLogger<T>, collectionName: string, schema: Schema) {
    this.db = new MultiTenantMongooseModel<T>(collectionName, schema);
    this.logHelper = logHelper;
  }

  async save(data: Partial<DataType<T>>, query?: any) {
    if (await this.documentExists(data)) {
      const saved = await this.db.findOneAndUpdate(
        query || { _id: data._id },
        data as UwaziUpdateQuery<DataType<T>>,
        {
          new: true,
        }
      );
      if (saved === null) {
        throw Error('The document was not updated!');
      }
      await this.logHelper.upsertLogOne(saved);
      return saved.toObject<WithId<T>>();
    }
    const saved = await this.db.create(data);
    await this.logHelper.upsertLogOne(saved);
    return saved.toObject<WithId<T>>();
  }

  async saveMultiple(data: Partial<DataType<T>>[]) {
    return Promise.all(data.map(async d => this.save(d)));
  }

  async updateMany(
    conditions: UwaziFilterQuery<DataType<T>>,
    doc: UwaziUpdateQuery<T>,
    options: ModelUpdateOptions = {}
  ) {
    await this.logHelper.upsertLogMany(conditions);
    return this.db._updateMany(conditions, doc, options);
  }

  async count(query: UwaziFilterQuery<DataType<T>> = {}) {
    return this.db.countDocuments(query);
  }

  async get(
    query: UwaziFilterQuery<DataType<T>> = {},
    select: any = '',
    options: UwaziQueryOptions = {}
  ) {
    const results = await this.db.find(query, select, { lean: true, ...options });
    return results as EnforcedWithId<T>[];
  }

  async getById(id: any, select?: any) {
    const results = await this.db.findById(id, select);
    return results as EnforcedWithId<T> | null;
  }

  async delete(condition: any) {
    let cond = condition;
    if (mongoose.Types.ObjectId.isValid(condition)) {
      cond = { _id: condition };
    }
    await this.logHelper.upsertLogMany(cond, true);
    return this.db.deleteMany(cond);
  }
}

// models are accessed in api/sync, which cannot be type-safe since the document
// type is a request parameter. Thus, we store all OdmModels as type Document.
export const models: { [index: string]: OdmModel<any> } = {};

export function instanceModel<T = any>(collectionName: string, schema: Schema) {
  const logHelper = createUpdateLogHelper<T>(collectionName);
  const model = new OdmModel<T>(logHelper, collectionName, schema);
  models[collectionName] = model;
  return model;
}
