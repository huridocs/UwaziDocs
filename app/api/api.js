import activitylogMiddleware from './activitylog/activitylogMiddleware';
import CSRFMiddleware from './auth/CSRFMiddleware';
import languageMiddleware from './utils/languageMiddleware';
import { setupApiSockets } from './socketio/setupSockets';
import { documentRoutes } from './documents/routes';
import { searchRoutes } from './search.v2/routes';
import { ocrRoutes } from './files/ocrRoutes';
import { permissionRoutes } from './permissions/routes';
import { suggestionsRoutes } from './suggestions/routes';
import { extractorsRoutes } from './suggestions/extractorsRoutes';
import { PreserveRoutes } from './preserve/routes';
import auth2fa from './auth2fa/routes';
import relationships from './relationships/routes';
import relationshipsV2 from './relationships.v2/routes/routes';
import activitylog from './activitylog/routes';
import users from './users/routes';
import templates from './templates/routes';
import searchDeprecatedRoutes from './search/deprecatedRoutes';
import search from './search/routes';
import topicclassification from './topicclassification/routes';
import thesauri from './thesauri/routes';
import relationtypes from './relationtypes/routes';
import documents from './documents/deprecatedRoutes';
import contact from './contact/routes';
import entities from './entities/routes';
import pages from './pages/routes';
import filesJSRoutes from './files/jsRoutes.js';
import files from './files/routes';
import filesExportRouter from './files/exportRoutes';
import settings from './settings/routes';
import i18n from './i18n/routes';
import sync from './sync/routes';
import tasks from './tasks/routes';
import usergroups from './usergroups/routes';
import stats from './stats/routes';

export default (app, server) => {
  app.use(CSRFMiddleware);
  app.use(languageMiddleware);
  app.use(activitylogMiddleware);
  setupApiSockets(server, app);
  documentRoutes(app);
  searchRoutes(app);
  ocrRoutes(app);
  permissionRoutes(app);
  suggestionsRoutes(app);
  extractorsRoutes(app);
  PreserveRoutes(app);
  auth2fa(app);
  relationships(app);
  relationshipsV2(app);
  activitylog(app);
  users(app);
  templates(app);
  searchDeprecatedRoutes(app);
  search(app);
  topicclassification(app);
  thesauri(app);
  relationtypes(app);
  documents(app);
  contact(app);
  entities(app);
  pages(app);
  filesJSRoutes(app);
  files(app);
  filesExportRouter(app);
  settings(app);
  i18n(app);
  sync(app);
  tasks(app);
  usergroups(app);
  stats(app);
};
