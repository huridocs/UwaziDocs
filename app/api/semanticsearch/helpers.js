import { entityDefaultDocument } from 'shared/entityDefaultDocument';
import { files } from 'api/files';

import { search } from '../search';
import model from './model';
import resultsModel from './resultsModel';
import { COMPLETED } from './statuses';
import templates from '../templates';

export const getSearchDocuments = async ({ documents, query }, language, user) => {
  if (documents && documents.length) {
    return documents;
  }
  const _query = { ...query, limit: 9999, searchTerm: '' };
  const res = await search.search(_query, language, user);
  return res.rows.map(doc => doc.sharedId);
};

export const removePageAnnotations = text => text.replace(/\[\[\d+\]\]/g, '');

export const updateSearchDocumentStatus = async (searchId, sharedId, status) =>
  model.db.findOneAndUpdate(
    {
      _id: searchId,
      'documents.sharedId': sharedId,
    },
    {
      $set: { 'documents.$.status': status },
    },
    { new: true, lean: true }
  );

export const setSearchDocumentResults = async (searchId, sharedId, results) => {
  const averageScore = results.length
    ? results.reduce((total, curr) => total + curr.score, 0) / results.length
    : 0;
  const _results = [...results];
  const docResults = await resultsModel.db.findOneAndUpdate(
    {
      sharedId,
      searchId,
    },
    {
      sharedId,
      searchId,
      averageScore,
      results: _results.sort((r1, r2) => r2.score - r1.score),
      status: COMPLETED,
    },
    { upsert: true, new: true }
  );
  return docResults;
};

export const extractDocumentContent = async (
  { documents, language, template, metadata },
  defaultLanguage
) => {
  const contents = {};
  let fullText;
  const defaultDocument = entityDefaultDocument(documents, language, defaultLanguage);
  if (defaultDocument && defaultDocument._id) {
    [{ fullText }] = await files.get(defaultDocument, '+fullText');
  }

  if (metadata) {
    (await templates.getById(template)).properties
      .filter(prop => prop.type === 'markdown')
      .forEach(field => {
        if (metadata[field.name]) {
          contents[field.name] = metadata[field.name];
        }
      });
  }

  if (fullText) {
    Object.keys(fullText).forEach(page => {
      contents[page] = removePageAnnotations(fullText[page]);
    });
  }

  return contents;
};
