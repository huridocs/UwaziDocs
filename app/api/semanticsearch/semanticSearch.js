import { promisify } from 'util';
import async from 'async';
import { Types } from 'mongoose';
import settings from 'api/settings';
import date from 'api/utils/date.js';
import model from './model';
import resultsModel from './resultsModel';
import api from './api';
import documentsModel from '../documents';
import workers from './workerManager';
import { createError } from '../utils';
import { PENDING, COMPLETED, PROCESSING, IN_PROGRESS, STOPPED } from './statuses';
import {
  getSearchDocuments,
  updateSearchDocumentStatus,
  setSearchDocumentResults,
  extractDocumentContent,
} from './helpers';

const SEARCH_BATCH_SIZE = 5;

const eachLimitAsync = promisify(async.eachLimit);

const processDocument = async (searchId, searchTerm, sharedId, language) => {
  const [doc] = await documentsModel.get({ sharedId, language });
  const defaultLanguage = (await settings.get()).languages.find(l => l.default);

  await updateSearchDocumentStatus(searchId, sharedId, PROCESSING);
  const contents = await extractDocumentContent(doc, defaultLanguage);
  if (!Object.keys(contents).length) {
    return updateSearchDocumentStatus(searchId, sharedId, COMPLETED);
  }

  const results = await api.processDocument({
    searchTerm,
    contents,
  });
  const savedResults = await setSearchDocumentResults(searchId, sharedId, results);
  await updateSearchDocumentStatus(searchId, sharedId, COMPLETED);
  return savedResults;
};

const processSearchLimit = async (searchId, docLimit) => {
  const searchObject = await model.save({ _id: searchId, status: IN_PROGRESS });
  const { language, searchTerm } = searchObject;
  const docs = searchObject.documents.filter(doc => doc.status !== COMPLETED);
  const docsToSearch = docs.length > docLimit ? docs.slice(0, docLimit) : docs;
  await eachLimitAsync(docsToSearch, SEARCH_BATCH_SIZE, async doc =>
    processDocument(searchId, searchTerm, doc.sharedId, language)
  );
  const updatedSearch = await model.getById(searchId);
  const isNotDone = updatedSearch.documents.some(doc => doc.status !== COMPLETED);
  if (isNotDone) {
    return { updatedSearch, processedDocuments: docsToSearch.map(d => d.sharedId) };
  }
  return {
    updatedSearch: await model.save({ _id: searchId, status: COMPLETED }),
    processedDocuments: docsToSearch.map(d => d.sharedId),
  };
};

const create = async (args, language, user) => {
  const docs = await getSearchDocuments(args, language, user);
  const newSearch = {
    documents: docs.map(docId => ({
      sharedId: docId,
      status: PENDING,
    })),
    status: PENDING,
    searchTerm: args.searchTerm,
    query: args.query,
    language,
    creationDate: date.currentUTC(),
  };
  const savedSearch = await model.save(newSearch);
  workers.notifyNewSearch(savedSearch._id);
  return savedSearch;
};

const getSearchResults = async (
  searchId,
  { skip = 0, limit = 30, threshold = 0.4, minRelevantSentences = 5 }
) =>
  resultsModel.db.aggregate([
    {
      $match: { searchId: Types.ObjectId(searchId) },
    },
    {
      $project: {
        searchId: 1,
        sharedId: 1,
        status: 1,
        results: 1,
        totalResults: { $size: '$results' },
        numRelevant: {
          $size: {
            $filter: {
              input: '$results',
              as: 'result',
              cond: { $gte: ['$$result.score', threshold] },
            },
          },
        },
      },
    },
    {
      $project: {
        totalResults: 1,
        searchId: 1,
        sharedId: 1,
        status: 1,
        results: 1,
        numRelevant: 1,
        relevantRate: { $divide: ['$numRelevant', '$totalResults'] },
      },
    },
    {
      $match: { numRelevant: { $gte: minRelevantSentences } },
    },
    {
      $sort: { relevantRate: -1 },
    },
    {
      $skip: skip,
    },
    {
      $limit: limit,
    },
  ]);

const getSearch = async (searchId, args) => {
  const theSearch = await model.getById(searchId);
  if (!theSearch) {
    throw createError('Search not found', 404);
  }

  const results = await getSearchResults(searchId, args);
  const docIds = results.map(r => r.sharedId);
  const docs = await documentsModel.get({
    sharedId: { $in: docIds },
    language: theSearch.language,
  });
  const docsWithResults = results.map(result => ({
    ...docs.find(doc => doc.sharedId === result.sharedId),
    semanticSearch: result,
  }));
  theSearch.results = docsWithResults;
  return theSearch;
};

const listSearchResultsDocs = async (searchId, args) => {
  const theSearch = await model.getById(searchId);
  if (!theSearch) {
    throw createError('Search not found', 404);
  }

  const { threshold, minRelevantSentences } = args;
  return resultsModel.db.aggregate([
    { $match: { searchId: Types.ObjectId(searchId) } },
    {
      $project: {
        sharedId: 1,
        numRelevant: {
          $size: {
            $filter: {
              input: '$results',
              as: 'result',
              cond: { $gte: ['$$result.score', threshold] },
            },
          },
        },
      },
    },
    { $match: { numRelevant: { $gte: minRelevantSentences } } },
    {
      $lookup: {
        from: 'entities',
        localField: 'sharedId',
        foreignField: 'sharedId',
        as: 'document',
      },
    },
    { $unwind: '$document' },
    { $match: { 'document.language': theSearch.language } },
    { $project: { _id: 0, sharedId: 1, template: '$document.template' } },
  ]);
};

const getDocumentResultsByIds = async (searchId, docIds) => {
  const theSearch = searchId._id ? searchId : await model.getById(searchId);
  const results = await resultsModel.get({ searchId, sharedId: { $in: docIds } });
  const docs = await documentsModel.get({
    sharedId: { $in: docIds },
    language: theSearch.language,
  });
  const docsWithResults = docs.map(doc => ({
    ...doc,
    semanticSearch: results.find(res => res.sharedId === doc.sharedId),
  }));
  return docsWithResults;
};

const getSearchesByDocument = async docId => {
  const results = await resultsModel.get({ sharedId: docId });
  const searchPromises = results.map(({ searchId }) => model.getById(searchId));
  const searches = await Promise.all(searchPromises);
  searches.map((theSearch, index) => ({
    ...theSearch,
    results: results[index],
  }));
  return searches;
};

const deleteSearch = async searchId => {
  const res = await model.delete(searchId);
  if (res.n !== 1) {
    throw createError('Search not found', 404);
  }
  await resultsModel.delete({ searchId });
  return { deleted: true };
};

const stopSearch = async searchId => {
  const res = await model.db.updateOne(
    {
      _id: searchId,
      status: { $in: [IN_PROGRESS, PENDING] },
    },
    {
      $set: { status: STOPPED },
    }
  );
  if (!res.n) {
    throw createError('Only running searches can be stopped', 400);
  }
  return model.getById(searchId);
};

const resumeSearch = async searchId => {
  const res = await model.db.updateOne(
    {
      _id: searchId,
      status: { $in: [STOPPED] },
    },
    {
      $set: { status: PENDING },
    }
  );
  if (!res.n) {
    throw createError('Only stopped searches can be resumed', 400);
  }
  workers.notifyNewSearch(searchId);
  return model.getById(searchId);
};

const getAllDocumentResults = async searchId => resultsModel.get({ searchId });
const getAllSearches = () => model.get().sort('-creationDate');
const getInProgress = async () => model.get({ status: IN_PROGRESS }).sort('-creationDate');
const getPending = async () => model.get({ status: PENDING }).sort('-creationDate');

const semanticSearch = {
  _id: Math.random(),
  create,
  processDocument,
  processSearchLimit,
  getAllDocumentResults,
  getDocumentResultsByIds,
  getAllSearches,
  getPending,
  getInProgress,
  getSearchResults,
  getSearch,
  listSearchResultsDocs,
  getSearchesByDocument,
  deleteSearch,
  stopSearch,
  resumeSearch,
};

export default semanticSearch;
