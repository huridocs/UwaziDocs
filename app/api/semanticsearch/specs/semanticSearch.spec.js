import db from 'api/utils/testing_db';
import date from 'api/utils/date.js';
import workers from '../workerManager';
import semanticSearch from '../semanticSearch';
import model from '../model';
import resultsModel from '../resultsModel';
import api from '../api';
import * as helpers from '../helpers';

import fixtures, {
  search1Id,
  search2Id,
  search3Id,
  searchIdForFilters,
  doc1Id,
  docWithoutTextId,
} from './fixtures';
import { createError } from '../../utils';

describe('semanticSearch', () => {
  beforeEach(async () => {
    await db.clearAllAndLoad(fixtures);
  });
  afterAll(async () => {
    await db.disconnect();
  });

  describe('create', () => {
    beforeEach(() => {
      jest.spyOn(workers, 'notifyNewSearch').mockImplementation(() => {});
      jest.spyOn(helpers, 'getSearchDocuments');
    });
    afterEach(() => {
      workers.notifyNewSearch.mockClear();
    });
    it('should create and save new search', async () => {
      jest.spyOn(date, 'currentUTC').mockReturnValue(1000);
      const args = { searchTerm: 'Test term', documents: ['doc1', 'doc2'] };
      const created = await semanticSearch.create(args, 'en', 'user');
      const savedSearch = await model.getById(created._id);
      delete savedSearch._id;
      expect(helpers.getSearchDocuments).toHaveBeenCalledWith(args, 'en', 'user');
      expect(savedSearch).toMatchSnapshot();
    });
    it('should send the search to the workers', async () => {
      const args = { searchTerm: 'Test term', documents: ['doc1', 'doc2'] };
      const created = await semanticSearch.create(args, 'en', 'user');
      expect(workers.notifyNewSearch).toHaveBeenCalledWith(created._id);
    });
  });

  describe('processDocument', () => {
    let expectedResults;
    beforeEach(() => {
      expectedResults = [
        { page: 1, text: 'page 1', score: 0.2 },
        { page: 2, text: 'page 2', score: 0.6 },
      ];
      jest.spyOn(api, 'processDocument').mockResolvedValue(expectedResults);
      api.processDocument.mockClear();
    });

    it('should send document to semantic search api for processing', async () => {
      await semanticSearch.processDocument(search1Id, 'legal', doc1Id, 'en');
      expect(api.processDocument).toHaveBeenCalledWith({
        searchTerm: 'legal',
        contents: {
          1: 'page 1',
          2: 'page 2',
        },
      });
    });

    it('should update the status of the document to be completed', async () => {
      await semanticSearch.processDocument(search1Id, 'legal', doc1Id, 'en');
      const theSearch = await model.getById(search1Id);
      const docInSearch = theSearch.documents.find(doc => doc.sharedId === doc1Id);
      expect(docInSearch.status).toBe('completed');
    });

    it('should save the results of the document sorted by score in descending order, and compute average score', async () => {
      await semanticSearch.processDocument(search1Id, 'legal', doc1Id, 'en');
      const [docResults] = await resultsModel.get({ searchId: search1Id, sharedId: doc1Id });
      expect(docResults.status).toBe('completed');
      expect(docResults.averageScore).toBe((0.2 + 0.6) / 2);
      expect(
        docResults.results.map(({ page, text, score }) => ({ page, text, score }))
      ).toMatchSnapshot();
    });

    describe('if document has no fullText or rich text fields', () => {
      it('should mark as completed without processing', async () => {
        jest.spyOn(helpers, 'extractDocumentContent').mockResolvedValue({});
        await semanticSearch.processDocument(search1Id, 'legal', docWithoutTextId, 'en');
        expect(api.processDocument).not.toHaveBeenCalled();
        const theSearch = await model.getById(search1Id);
        const docInSearch = theSearch.documents.find(doc => doc.sharedId === docWithoutTextId);
        expect(docInSearch.status).toBe('completed');
        const docRes = await resultsModel.get({ searchId: search1Id, sharedId: docWithoutTextId });
        expect(docRes.length).toBe(0);
      });
    });
  });

  describe('processSearchLimit', () => {
    const expectedResults = [
      { page: '1', text: 'page 1', score: 0.6 },
      { page: '2', text: 'page 2', score: 0.2 },
    ];
    beforeEach(() => {
      jest.spyOn(api, 'processDocument').mockResolvedValue(expectedResults);
      api.processDocument.mockClear();
    });

    it('should process only up to specified number of unprocessed docs in the search', async () => {
      helpers.extractDocumentContent.mockRestore();
      await semanticSearch.processSearchLimit(search2Id, 2);
      expect(api.processDocument).toHaveBeenCalledTimes(2);
      expect(api.processDocument).toHaveBeenCalledWith({
        searchTerm: 'injustice',
        contents: { 1: 'text2' },
      });
      expect(api.processDocument).toHaveBeenCalledWith({
        searchTerm: 'injustice',
        contents: { 1: 'text3' },
      });
      const theSearch = await model.getById(search2Id);
      expect(
        theSearch.documents.some(doc => doc.sharedId === 'doc2' && doc.status === 'completed')
      ).toBe(true);
      expect(
        theSearch.documents.some(doc => doc.sharedId === 'doc3' && doc.status === 'completed')
      ).toBe(true);
      expect(
        theSearch.documents.some(doc => doc.sharedId === 'doc5' && doc.status === 'pending')
      ).toBe(true);
      const res = await resultsModel.get({ searchId: search2Id });
      expect(res.some(r => r.sharedId === 'doc2')).toBe(true);
      expect(res.some(r => r.sharedId === 'doc3')).toBe(true);

      expect(theSearch.status).toBe('inProgress');
    });
    it('should return the updated search and processed documents', async () => {
      const res = await semanticSearch.processSearchLimit(search2Id, 2);
      expect(res.updatedSearch._id).toEqual(search2Id);
      delete res.updatedSearch._id;
      expect(res).toMatchSnapshot();
    });
    it('should mark search as complete if all documents are processed', async () => {
      await semanticSearch.processSearchLimit(search2Id, 5);
      const theSearch = await model.getById(search2Id);
      expect(theSearch.status).toBe('completed');
    });
  });

  describe('getAllSearches', () => {
    it('should return all searches', async () => {
      const searches = await semanticSearch.getAllSearches();
      expect(searches[0]._id).toEqual(search1Id);
      expect(searches.length).toBe(4);
    });
  });

  describe('getPending', () => {
    it('should return pending searches', async () => {
      const pending = await semanticSearch.getPending();
      expect(pending.length).toBe(1);
      expect(pending[0]._id).toEqual(search2Id);
    });
  });

  describe('getInProgress', () => {
    it('should return all searches in progress', async () => {
      const inProgress = await semanticSearch.getInProgress();
      expect(inProgress.length).toBe(1);
      expect(inProgress[0]._id).toEqual(search1Id);
    });
  });

  describe('getAllDocumentResults', () => {
    it('should return all document results of the specified search', async () => {
      const results = await semanticSearch.getAllDocumentResults(search3Id);
      expect(results.length).toEqual(3);
      expect(results.some(doc => doc.sharedId === 'doc1')).toBe(true);
      expect(results.some(doc => doc.sharedId === 'doc2')).toBe(true);
      expect(results.some(doc => doc.sharedId === 'doc3')).toBe(true);
    });
  });

  describe('getSearchResults', () => {
    it('should return search results filtered with the specified args and sorted by proportion of relevant docs', async () => {
      const args = {
        skip: 1,
        limit: 2,
        minRelevantSentences: 1,
        threshold: 0.6,
      };
      const results = await semanticSearch.getSearchResults(searchIdForFilters, args);
      expect(results.map(r => r.sharedId)).toEqual(['3', '2']);
      expect(
        results.map(r => {
          const withoutIds = { ...r };
          delete withoutIds._id;
          delete withoutIds.searchId;
          return withoutIds;
        })
      ).toMatchSnapshot();
    });
  });

  describe('getSearch', () => {
    it('should fetch a search by id and its document entities with filtered semantic search results', async () => {
      const res = await semanticSearch.getSearch(search3Id, {
        threshold: 0.5,
        minRelevantSentences: 1,
      });
      res.results.forEach(doc => {
        //eslint-disable-next-line no-param-reassign
        delete doc._id;
        //eslint-disable-next-line no-param-reassign
        delete doc.semanticSearch._id;
        //eslint-disable-next-line no-param-reassign
        delete doc.semanticSearch.searchId;
        //eslint-disable-next-line no-param-reassign
        delete doc.documents[0]._id;
      });
      delete res._id;
      expect(res.results.length).toBe(2);
      expect(res).toMatchSnapshot();
    });
    it('should return 404 if search does not exist', async () => {
      try {
        await semanticSearch.getSearch(db.id(), {});
        fail('should throw error');
      } catch (e) {
        expect(e).toEqual(createError('Search not found', 404));
      }
    });
  });

  describe('listSearchResultsDocs', () => {
    it('should return the shared id and templates of all results that match the filters', async () => {
      const args = {
        minRelevantSentences: 2,
        threshold: 0.6,
      };
      const results = await semanticSearch.listSearchResultsDocs(searchIdForFilters, args);
      const docIds = results.map(r => r.sharedId);
      expect(results.length).toBe(3);
      expect(docIds.includes('2')).toBe(true);
      expect(docIds.includes('3')).toBe(true);
      expect(docIds.includes('4')).toBe(true);
      expect(results).toMatchSnapshot();
    });
  });

  describe('deleteSearch', () => {
    it('should delete specified search and its results', async () => {
      await semanticSearch.deleteSearch(search3Id);
      const searchInDb = await model.getById(search3Id);
      expect(searchInDb).toBeNull();
      const searchResults = await resultsModel.get({ searchId: search3Id });
      expect(searchResults.length).toBe(0);
    });
    it('should return 404 error if search does not exist', async () => {
      try {
        await semanticSearch.deleteSearch(db.id());
        fail('should throw error');
      } catch (e) {
        expect(e).toEqual(createError('Search not found', 404));
      }
    });
  });

  describe('stopSearch', () => {
    const testErrorIfWrongStatus = async (searchId, status) => {
      try {
        await semanticSearch.stopSearch(searchId);
        fail('should throw error');
      } catch (e) {
        const theSearch = await model.getById(searchId);
        expect(theSearch.status).toBe(status);
        expect(e.code).toBe(400);
      }
    };
    it('should set status to stopped if search is pending', async () => {
      await semanticSearch.stopSearch(search2Id);
      const updatedSearch = await model.getById(search2Id);
      expect(updatedSearch.status).toBe('stopped');
    });
    it('should set status to stopped if search is in progress', async () => {
      await semanticSearch.stopSearch(search1Id);
      const updatedSearch = await model.getById(search1Id);
      expect(updatedSearch.status).toBe('stopped');
    });
    it('should return the updated search', async () => {
      const stoppedSearch = await semanticSearch.stopSearch(search1Id);
      expect(stoppedSearch.status).toBe('stopped');
    });
    it('should throw error if search is completed', async () => {
      await testErrorIfWrongStatus(search3Id, 'completed');
    });
    it('should throw error if search is stopped', async () => {
      await model.db.findOneAndUpdate({ _id: search3Id }, { $set: { status: 'stopped' } });
      await testErrorIfWrongStatus(search3Id, 'stopped');
    });
    it('should throw error if search does not exist', async () => {
      try {
        await semanticSearch.stopSearch(db.id());
        fail('should throw error');
      } catch (e) {
        expect(e.code).toBe(400);
      }
    });
  });

  describe('resumeSearch', () => {
    beforeEach(() => {
      jest.spyOn(workers, 'notifyNewSearch').mockImplementation(() => {});
    });
    const testErrorIfWrongStatus = async (searchId, status) => {
      await model.db.updateOne({ _id: searchId }, { $set: { status } });
      try {
        await semanticSearch.resumeSearch(searchId);
        fail('should throw error');
      } catch (e) {
        const theSearch = await model.getById(searchId);
        expect(theSearch.status).toBe(status);
        expect(e.code).toBe(400);
      }
    };
    it('should set status to pending if search is stopped', async () => {
      await model.db.updateOne({ _id: search1Id }, { $set: { status: 'stopped' } });
      await semanticSearch.resumeSearch(search1Id);
      const resumedSearch = await model.getById(search1Id);
      expect(resumedSearch.status).toBe('pending');
    });
    it('should notify workerManager', async () => {
      await model.db.updateOne({ _id: search1Id }, { $set: { status: 'stopped' } });
      await semanticSearch.resumeSearch(search1Id);
      expect(workers.notifyNewSearch).toHaveBeenCalledWith(search1Id);
    });
    it('should return the updated search', async () => {
      await model.db.updateOne({ _id: search1Id }, { $set: { status: 'stopped' } });
      const resumed = await semanticSearch.resumeSearch(search1Id);
      expect(resumed.status).toBe('pending');
    });
    it('should throw error if search is not stopped', async () => {
      const statuses = ['completed', 'pending', 'inProgress'];
      const searchIds = [search1Id, search2Id, search3Id];
      await Promise.all(
        statuses.map((status, index) => testErrorIfWrongStatus(searchIds[index], status))
      );
    });
  });
});
