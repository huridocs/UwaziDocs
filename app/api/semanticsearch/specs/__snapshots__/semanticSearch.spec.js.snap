// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`semanticSearch create should create and save new search 1`] = `
Object {
  "__v": 0,
  "creationDate": 1000,
  "documents": Array [
    Object {
      "sharedId": "doc1",
      "status": "pending",
    },
    Object {
      "sharedId": "doc2",
      "status": "pending",
    },
  ],
  "language": "en",
  "searchTerm": "Test term",
  "status": "pending",
}
`;

exports[`semanticSearch getSearch should fetch a search by id and its document entities with filtered semantic search results 1`] = `
Object {
  "documents": Array [
    Object {
      "sharedId": "doc1",
      "status": "completed",
    },
    Object {
      "sharedId": "doc2",
      "status": "completed",
    },
  ],
  "language": "en",
  "results": Array [
    Object {
      "documents": Array [
        Object {
          "entity": "doc2",
          "language": "en",
          "type": "document",
        },
      ],
      "language": "en",
      "semanticSearch": Object {
        "numRelevant": 2,
        "relevantRate": 0.6666666666666666,
        "results": Array [
          Object {
            "page": 1,
            "score": 0.8,
            "sentence": "that",
          },
          Object {
            "page": 4,
            "score": 0.6,
            "sentence": "then",
          },
          Object {
            "page": 2,
            "score": 0.1,
            "sentence": "what",
          },
        ],
        "sharedId": "doc2",
        "totalResults": 3,
      },
      "sharedId": "doc2",
    },
    Object {
      "documents": Array [
        Object {
          "entity": "doc1",
          "language": "en",
          "type": "document",
        },
      ],
      "language": "en",
      "semanticSearch": Object {
        "numRelevant": 1,
        "relevantRate": 0.5,
        "results": Array [
          Object {
            "page": 1,
            "score": 0.5,
            "sentence": "stuff",
          },
          Object {
            "page": 2,
            "score": 0.2,
            "sentence": "this",
          },
        ],
        "sharedId": "doc1",
        "totalResults": 2,
      },
      "sharedId": "doc1",
    },
  ],
  "searchTerm": "torture",
  "status": "completed",
}
`;

exports[`semanticSearch getSearchResults should return search results filtered with the specified args and sorted by proportion of relevant docs 1`] = `
Array [
  Object {
    "numRelevant": 3,
    "relevantRate": 0.75,
    "results": Array [
      Object {
        "score": 0.8,
      },
      Object {
        "score": 0.2,
      },
      Object {
        "score": 0.67,
      },
      Object {
        "score": 0.71,
      },
    ],
    "sharedId": "3",
    "status": "completed",
    "totalResults": 4,
  },
  Object {
    "numRelevant": 2,
    "relevantRate": 0.6666666666666666,
    "results": Array [
      Object {
        "score": 0.7,
      },
      Object {
        "score": 0.9,
      },
      Object {
        "score": 0.3,
      },
    ],
    "sharedId": "2",
    "status": "completed",
    "totalResults": 3,
  },
]
`;

exports[`semanticSearch listSearchResultsDocs should return the shared id and templates of all results that match the filters 1`] = `
Array [
  Object {
    "sharedId": "2",
    "template": "t2",
  },
  Object {
    "sharedId": "3",
    "template": "t3",
  },
  Object {
    "sharedId": "4",
    "template": "t4",
  },
]
`;

exports[`semanticSearch processDocument should save the results of the document sorted by score in descending order, and compute average score 1`] = `
Array [
  Object {
    "page": "2",
    "score": 0.6,
    "text": "page 2",
  },
  Object {
    "page": "1",
    "score": 0.2,
    "text": "page 1",
  },
]
`;

exports[`semanticSearch processSearchLimit should return the updated search and processed documents 1`] = `
Object {
  "processedDocuments": Array [
    "doc2",
    "doc3",
  ],
  "updatedSearch": Object {
    "documents": Array [
      Object {
        "sharedId": "doc1",
        "status": "completed",
      },
      Object {
        "sharedId": "doc2",
        "status": "completed",
      },
      Object {
        "sharedId": "doc3",
        "status": "completed",
      },
      Object {
        "sharedId": "doc4",
        "status": "completed",
      },
      Object {
        "sharedId": "doc5",
        "status": "pending",
      },
    ],
    "language": "en",
    "searchTerm": "injustice",
    "status": "inProgress",
  },
}
`;
